### PR TITLE
Fastq Alignment Parallelization 

### PR DESCRIPTION
**Pull request by trevormartinj7**
This project splits the up the work of creating multiple variant objects for .fastq files across multiple processes, greatly speeding up processing time for processing of all .fastq files.

It uses python's multiprocessing library to spin up processes which then create variant objects in parallel. It then stores them in intermediate .tsv files, which are read to track statistics of the files.

### CODE REVIEW

File: CRISPResso2/CRISPRessoCORE.py [hunk 1]
```diff
@@ -514,95 +575,151 @@ def process_fastq(fastq_filename, variantCache, ref_names, refs, args):
...
+                if READ_LENGTH == 0:
+                    READ_LENGTH = len(variant[match_name]['aln_seq'])
+                N_GLOBAL_SUBS += (variant[match_name]['substitution_n'] + variant[match_name]['substitutions_outside_window']) * variant_count
+                N_SUBS_OUTSIDE_WINDOW += variant[match_name]['substitutions_outside_window'] * variant_count
+                N_MODS_IN_WINDOW += variant[match_name]['mods_in_window'] * variant_count
+                N_MODS_OUTSIDE_WINDOW += variant[match_name]['mods_outside_window'] * variant_count
+                if variant[match_name]['irregular_ends']:
+                    N_READS_IRREGULAR_ENDS += variant_count
+            if (index % 10000 == 0):
+                info("Processing Reads; %d Completed out of %d Unique Reads"%(index, num_unique_reads))
```
Comment (Colelyman) [inline CRISPResso2/CRISPRessoCORE.py:704]: Very minor, but I think `Processing reads; %d completed out of %d unique reads` may look more inline with the other info statements.

File: CRISPResso2/CRISPRessoCORE.py [hunk 2]
```diff
@@ -2716,6 +2834,9 @@ def get_allele_row(reference_name, variant_count, aln_ref_names_str, aln_ref_sco
         ###iterate through variants
         for variant in variantCache:
             #skip variant if there were none observed
+            # check if variant is an int
+            if isinstance(variantCache[variant], int):
+                print(variant)
```
Comment (Colelyman) [inline CRISPResso2/CRISPRessoCORE.py:2839]: Is this needed?

**Review by Colelyman — CHANGES_REQUESTED** [review]
This is looking really good! Thanks for all of your work. I just have a few minor (mostly formatting) comments.

File: CRISPResso2/CRISPRessoCORE.py [hunk 3]
```diff
@@ -441,7 +442,67 @@ def get_new_variant_object(args, fastq_seq, refs, ref_names, aln_matrix, pe_scaf
...
+        
+    """
+    new_variants = {}
+    num_processed = 0
+    for fastq_seq in seq_list:
+        new_variant = get_new_variant_object(args, fastq_seq, refs, ref_names, aln_matrix, pe_scaffold_dna_info)
+        new_variants[fastq_seq] = new_variant
+        num_processed += 1
+        if num_processed % 1000 == 0:
+            info(f"Process {process_id + 1} has processed {num_processed} unique reads")
```
[the remark this answers is not in the document]
Comment (trevormartinj7) [reply CRISPResso2/CRISPRessoCORE.py:499]: Sounds good to me, will result in less info clutter for medium sized runs

File: CRISPResso2/CRISPRessoCORE.py [hunk 4]
```diff
@@ -514,95 +575,151 @@ def process_fastq(fastq_filename, variantCache, ref_names, refs, args):
...
+    N_MODS_IN_WINDOW = 0 # number of modifications found inside the quantification window
+    N_MODS_OUTSIDE_WINDOW = 0 # number of modifications found outside the quantification window
+    N_READS_IRREGULAR_ENDS = 0 # number of reads with modifications at the 0 or -1 position
+    READ_LENGTH = 0
+    unaligned_reads = []
+
+    if n_processes > 1:
+        boundaries = get_variant_cache_boundaries(num_unique_reads, n_processes)
+
+        managerCache = Manager().dict() # Manager dictionaries are python multiprocessing objects that can be accessed across processes
```
[the remark this answers is not in the document]
Comment (trevormartinj7) [reply CRISPResso2/CRISPRessoCORE.py:637]: Done. I had it the other way to match variantCache, but snakecase fits other variables better

File: CRISPResso2/CRISPRessoCORE.py (hunk 2 above)
Comment (trevormartinj7) [reply CRISPResso2/CRISPRessoCORE.py:2839]: No, that's code that I used to debug the empty string, removed

File: CRISPResso2/CRISPRessoCORE.py [hunk 5]
```diff
@@ -441,7 +442,67 @@ def get_new_variant_object(args, fastq_seq, refs, ref_names, aln_matrix, pe_scaf
 
     return new_variant
 
+
+
+def get_variant_cache_boundaries(num_unique_sequences, n_processes):
+    """get_variant_cache_boundaries determines the boundaries for the number of unique sequences to be processed by each process
+        num_unique_sequences: the number of unique sequences to be processed
+        n_processes: the number of processes to be used
+    """
```
[the remark this answers is not in the document]
Comment (trevormartinj7) [reply CRISPResso2/CRISPRessoCORE.py:451]: Done!

**Pull request closed without merging**